### ISSUE TITLE
[HOLD for payment 2022-12-22][$500] Bug: Old dot link without ‘www.’ prefix ( expensify.com/inbox) does not auto signin to old dot reported by @sobitneupane

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#11810]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:

1. Login to new dot expensify.
2. Send old dot link without ‘www.’ prefix (example: [expensify.com/inbox](http://expensify.com/inbox))
3. Press the link.

## Expected Result:
You are automatically signed in to old dot.

## Actual Result:
You are prompted to enter password.

## Workaround:
unknown
## Platform:

Where is this issue occurring?

- Web
- iOS
- Android
- Desktop App
- Mobile Web

**Version Number:** 1.2.14
**Reproducible in staging?:** y
**Reproducible in production?:** y
**Email or phone of affected tester (no customers):** 
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** 

https://user-images.githubusercontent.com/43996225/195652418-0de43661-52fa-4eab-9290-7553df8d2252.mov

**Expensify/Expensify Issue URL:** 
**Issue reported by:** @sobitneupane
**Slack conversation:** https://expensify.slack.com/archives/C01GTK53T8Q/p1665568331616469

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

https://user-images.githubusercontent.com/43996225/195652444-18c7006d-2fae-4355-8b01-6c5ab9894113.mp4

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~011559e465b3832421</li>
        <li>Upwork Job ID: 1603853910330884096</li>
        <li>Last Price Increase: 2022-12-23</li>
    </ul>
</details>

**Comment by conorpendergrast** [comment]
Didn't get to this before I went OoO, reassigning!

**Comment by sakluger** [comment]
Reproduced in prod in an incognito window.

**Comment by MonilBhavsar** [comment]
Can be worked externally

**Comment by tienifr** [comment]
# Proposal
## Problem
In https://github.com/Expensify/App/blob/d6baba0be2da93beee72d8f21a6ee83e297fdd1a/src/components/HTMLEngineProvider/HTMLRenderers/AnchorRenderer.js#L28-L48
we check `attrHref.startsWith(CONFIG.EXPENSIFY.EXPENSIFY_URL)` to detect it's expensify url or not, but `EXPENSIFY_URL` = `https://www.expensify.com/`

## Solution
In https://github.com/Expensify/App/blob/d6baba0be2da93beee72d8f21a6ee83e297fdd1a/src/components/HTMLEngineProvider/HTMLRenderers/AnchorRenderer.js
We'll remove `www.` in `attrHref` and `expensifyUrl` before we check `attrHref.startsWith(CONFIG.EXPENSIFY.EXPENSIFY_URL)`.

```diff

+    const attrHrefWithoutW3 = attrHref.startsWith('www.')? attrHref.replace('www.',''):attrHref;
+    const expensifyUrlWithoutW3 = CONFIG.EXPENSIFY.EXPENSIFY_URL.replace('www.','');
+    const conciergeUrlWithoutW3 = CONFIG.EXPENSIFY.CONCIERGE_URL.replace('www.','');

    const internalNewExpensifyPath = (attrHref.startsWith(CONST.NEW_EXPENSIFY_URL) && attrHref.replace(CONST.NEW_EXPENSIFY_URL, ''))
        || (attrHref.startsWith(CONST.STAGING_NEW_EXPENSIFY_URL) && attrHref.replace(CONST.STAGING_NEW_EXPENSIFY_URL, ''));
-    const internalExpensifyPath = attrHref.startsWith(CONFIG.EXPENSIFY.EXPENSIFY_URL)
-                                    && !attrHref.startsWith(CONFIG.EXPENSIFY.CONCIERGE_URL)
-                                    && attrHref.replace(CONFIG.EXPENSIFY.EXPENSIFY_URL, '');
+    const internalExpensifyPath = attrHrefWithoutW3.startsWith(expensifyUrlWithoutW3)
+                                    && !attrHrefWithoutW3.startsWith(conciergeUrlWithoutW3)
+                                   && attrHrefWithoutW3.replace(expensifyUrlWithoutW3, '');
```

https://user-images.githubusercontent.com/113963320/198528098-8234dfc1-f9b4-4c9a-a3e9-fd64741e0295.mp4

**Comment by parasharrajat** [comment]
Thanks for the proposal @tienifr Good catch. Let's improve your solution. It would be great to create a common function that matches any two domains. I think simple rules will be. 
1. Create a function that takes two params `A` and `B` with type url. 
2. Exclude the parts after domain segment from URLs.
2. Return true when domains A and B are exact matches.
3. Return true when domains A and B  and their subdomains are a match. (if no subdomain is passed use `www` or if `www` subdomain is passed, ignore it) 
3 Ignore protocol part `https or http`. 
3. Return false for all other cases.

**Comment by tienifr** [comment]
# Update Proposal
@parasharrajat I updated my solution in here: https://github.com/tienifr/App/pull/18. Pls help to check, thanks.

**Comment by tienifr** [comment]
@conorpendergrast Thanks for your review. I realized that new URL not work on native, so I changed to use url libs. Pls help check: https://github.com/tienifr/App/pull/18

**Comment by parasharrajat** [comment]
@MonilBhavsar Are we fine with package `url`? https://github.com/Expensify/App/issues/11810#issuecomment-1298209880
![image](https://user-images.githubusercontent.com/24370807/199211556-eee881ee-2994-4475-b145-bef1d570827b.png)

**Comment by MonilBhavsar** [comment]
Hmm, This looks good. But I'm not sure about the process of using new package. Let's post this in channel and get thoughts from the team?

**Comment by MonilBhavsar** [comment]
@parasharrajat let's discuss this in slack about adding and using a new package. Do you mind opening a conversation with the context you have?

**Comment by parasharrajat** [comment]
Asked in https://expensify.slack.com/archives/C01GTK53T8Q/p1667833896505429. Please chime in.

**Comment by parasharrajat** [comment]
@tienifr Is it possible to use regExp or something to get rid of `url` package from your solution?

**Comment by tienifr** [comment]
@parasharrajat I've updated my PR to use regExp: https://github.com/tienifr/App/pull/18. Pls help check again. Thanks

**Comment by parasharrajat** [comment]
const reURLInformation = new RegExp([
        '^(https?:)//', // protocol
        '(([^:/?#]*)(?::([0-9]+))?)', // host (hostname and port)
        '(/{0,1}[^?#]*)', // pathname
        '(\\?[^#]*|)', // search
        '(#.*|)$', // hash
    ].join(''));
    
I am not sure how good these are you will have to improve them during tests. try with different possible URLs. e.g `https://github.com/Expensify/App/issues/7853`

Also, the purpose of creating this function is to replace all usages of manual URL matching. Both `internalNewExpensifyPath` and `internalExpensifyPath` should be replaced with this. 

Also, a better name for the function would be `getURLObject`. We can drop the unnecessary parts e.g. `hash` `search` as they are not needed for this. 

Rest, @tienifr 's proposal looks good to me. 

cc: @MonilBhavsar 

:ribbon: :eyes: :ribbon:  C+ reviewed

**Comment by tienifr** [comment]
@parasharrajat Thanks for your review. I've updated my sample PR: https://github.com/tienifr/App/pull/18

**Comment by MonilBhavsar** [comment]
Proposal looks good to me 👍

**Comment by MonilBhavsar** [comment]
@tienifr left some comments on your proposal/PR after initial review

**Comment by tienifr** [comment]
Hi @parasharrajat  @MonilBhavsar  I just fixed your comments and created PR on the Expensify repo. Please help to review it. Thanks

**Comment by CortneyOfstad** [comment]
I just realized the UpWork job was not posted, so I finished that — https://www.upwork.com/jobs/~018c4bbda06b5c47e0. 

Was prompted to double it, so set it at $500.

**Comment by parasharrajat** [comment]
@CortneyOfstad we already [hire](https://github.com/Expensify/App/issues/11810#issuecomment-1311702342) a contributor for this job so it should not be doubled.

**Comment by CortneyOfstad** [comment]
Sorry about that! Changing it back now!

**Comment by parasharrajat** [comment]
In review.

**Comment by tienifr** [comment]
Thanks @CortneyOfstad, I've applied to the job

**Comment by CortneyOfstad** [comment]
@tienifr @parasharrajat @sobitneupane I sent over the contract details in UpWork. Let me know if you have any questions!

**Comment by CortneyOfstad** [comment]
Not overdue

**Comment by MonilBhavsar** [comment]
Were required to change expensify-common lib
That PR is merged, and expensify-common lib version is updated in App repo.
Now onto the final App PR

**Comment by MonilBhavsar** [comment]
Reviewing PR's

**Comment by MonilBhavsar** [comment]
Discussing in PR

**Comment by MonilBhavsar** [comment]
^^ Same, Discussing in PR

**Comment by CortneyOfstad** [comment]
@MonilBhavsar @parasharrajat @tienifr Sorry for the delay on this — I'm wondering about the payment. Does this need to be made at this point since a fix has not yet been implemented? TIA!

**Comment by parasharrajat** [comment]
No payments yet. Please reset the title and let's wait for the final PR to be merged. 

I would also request a bonus for @tienifr because it required working on multiple PRs and we asked them to create a more generalized solution.

**Comment by MonilBhavsar** [comment]
PR in review. Discussing there

**Comment by trjExpensify** [comment]
👋 Nice! [C+ approved, over to you Monil](https://github.com/Expensify/App/pull/12701#issuecomment-1347995633). This is an oldie, excited to get it over the line! 

Also, friendly reminder to label the issue as `Reviewing` when the linked PR is in review. Thanks!

**Comment by CortneyOfstad** [comment]
Heading OoO for two weeks starting now, so reassigning 👍

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (First week)

**Comment by MonilBhavsar** [comment]
Ah, the final PR was merged and has already passed a week since it was deployed.
We're fine to issue the payment and close this

**Comment by MonilBhavsar** [comment]
The solution for this issue has been :rocket: deployed to production :rocket: in version 1.2.40-3 and is now subject to a 7-day regression period :calendar:. Here is the list of pull requests that resolve this issue:

- https://github.com/Expensify/App/pull/12701


If no regressions arise, payment will be issued on 2022-12-22. :confetti_ball:

After the hold period, please check if any of the following need payment for this issue, and if so check them off after paying:
- [ ] External issue reporter
- [ ] Contributor that fixed the issue
- [ ] Contributor+ that helped on the issue and/or PR

As a reminder, here are the bonuses/penalties that should be applied for any External issue:
- Merged PR within 3 business days of assignment - 50% **bonus**
- Merged PR more than 9 business days after assignment - 50% **penalty**

**Comment by MonilBhavsar** [comment]
**BugZero Checklist:** The PR fixing this issue has been merged! The following checklist ([instructions](https://stackoverflow.com/c/expensify/questions/14574)) will need to be completed before the issue can be closed:

- [ ] [@MonilBhavsar @parasharrajat] The PR that introduced the bug has been identified. Link to the PR:
- [ ] [@MonilBhavsar @parasharrajat] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment:
- [ ] [@MonilBhavsar @parasharrajat] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion:
- [ ] [@sonialiap] A [regression test](https://expensify.testrail.io/index.php?/dashboard) has been added or updated so that the same bug will not reach production again. Link to the GH issue for creating the test here:

**Comment by MonilBhavsar** [comment]
@sonialiap can we please issue the payments and close this issue

**Comment by mallenexpensify** [comment]
`Reviewing` label kept this from going overdue.  @sonialiap can you please ensure all payments are made and also complete the regression test steps?

**Comment by sonialiap** [comment]
On it 🖖

**Comment by sonialiap** [comment]
Following this [SO](https://stackoverflowteams.com/c/expensify/questions/14574), posted regression test [proposal in slack ](https://expensify.slack.com/archives/C01SKUP7QR0/p1673918731715469)

**Comment by sonialiap** [comment]
Created new regression test `oldDot links clicked in newDot automatically log user into oldDot`

**Comment by sonialiap** [comment]
Regression test GH issue https://github.com/Expensify/Expensify/issues/256201

**Comment by sonialiap** [comment]
@parasharrajat can you help me identify which PR caused this issue?

**Comment by MonilBhavsar** [comment]
Not a regression, I think. We missed this case initially while implementing this

**Comment by MonilBhavsar** [comment]
@sonialiap is issuing payment

**Comment by sonialiap** [comment]
Offer sent to @sobitneupane for reporting issue

**Comment by sonialiap** [comment]
@tienifr please apply for this job:

https://www.upwork.com/jobs/~01af93adacf3d34f62

@parasharrajat I believe I found your profile on Upwork and invited you to the job, offer sent :)

**Comment by tienifr** [comment]
> I would also request a bonus for @tienifr because it required working on multiple PRs and we asked them to create a more generalized solution. cc: @MonilBhavsar

@MonilBhavsar @parasharrajat I'd like to confirm on the bonus mentioned [here](https://github.com/Expensify/App/issues/11810#issuecomment-1343168314)

This issue requires many PRs, a generalized solution, with 70+ lines of code added and a new ~250 lines test file.
https://github.com/Expensify/App/pull/12701
https://github.com/Expensify/expensify-common/pull/475

**Comment by sonialiap** [comment]
@tienifr thanks for bringing that up! Please apply for https://www.upwork.com/jobs/~01af93adacf3d34f62 and I'll include the bonus

**Comment by tienifr** [comment]
@sonialiap should we align on the bonus amount on this thread first, similar to in [here](https://github.com/Expensify/App/issues/10893#issuecomment-1399318803)?

@parasharrajat @MonilBhavsar what do you think?

**Comment by sonialiap** [comment]
@tienifr yes, good idea. Do you have thoughts of how much the bonus should be on this issue? 

(Rajat, I am going to hold of sending a bonus payment until we figure out the bonus amount)

**Comment by tienifr** [comment]
@sonialiap I think $500 bonus would be reasonable considering the amount of changes made are rather huge compared to a regular issue.

But I'm all good with any decision made by @parasharrajat here.

I've applied to the job, thanks.

**Comment by sonialiap** [comment]
Sound good @tienifr. @MonilBhavsar, can you please confirm via a comment that you're cool with this and I'll issue payment?

**Comment by MonilBhavsar** [comment]
> I think $250 is fine for the additional PR as 50% bonus.
>
> Adding tests and clean up is considered part of work. It may not be needed for some PRs but it doesn't count as additional work.

I agree with Rajat's reasoning here and $250 bonus

**Comment by MonilBhavsar** [comment]
@sonialiap can we please issue the payment and bonus here

**Comment by MonilBhavsar** [comment]
@sonialiap is ooo

**Comment by mallenexpensify** [comment]
Bonused Rajat $250.  Paid @sobitneupane $250 for reporting. 
@tienifr can you please accept the job and reply here once you have? You'll get paid $750 total
https://www.upwork.com/jobs/~01af93adacf3d34f62

**Comment by tienifr** [comment]
@mallenexpensify accepted, thanks!

**Comment by mallenexpensify** [comment]
Paid @tienifr $750 
I added this to [the testing guidelines](https://docs.google.com/document/d/1UaWlcieiwLnsKfeKw7iwLF7vgYRmdw_-K8OcY5aPBRU/edit#heading=h.qm8gncdaia0r) we're working on, since it doesn't make sense (to me) to have a test case for it. 

As for the other regression checklist items, I don't remember this ever working correctly before, does anyone?

**Comment by MonilBhavsar** [comment]
@mallenexpensify thanks! 

> since it doesn't make sense (to me) to have a test case for it.

I think we should have a test for this in chat messages list. We discussed here https://expensify.slack.com/archives/C01SKUP7QR0/p1673918731715469

Regarding regression, I believe it is like this since beginning, as we missed this case.

**Comment by mallenexpensify** [comment]
oooh, I didn't know we had an area for `OldDot Transition` in TestRail, that sounds like the best place for it.  Thanks @MonilBhavsar

**Comment by sonialiap** [comment]
Thanks @mallenexpensify!

**Comment by MonilBhavsar** [comment]
Looks like everyone are paid.
we can close this once we have requested to add this to regression testing. cc @sonialiap

**Comment by sonialiap** [comment]
Was just following up on the [regression GH](https://github.com/Expensify/Expensify/issues/256201). Looks like all is set on this one, everyone's paid and the regression test is set up. We can close 👍

**Comment by MonilBhavsar** [comment]
Thanks!
